File: Ocis.Client/sdk/py/protocol.py
from __future__ import annotations
from typing import (Any, Generic, TypeVar)
from .Ocis_Server.protocol_spec import (RequestPacket, ResponsePacket)
from .fable_modules.fable_library.array_ import (get_sub_array, copy_to)
from .fable_modules.fable_library.encoding import get_utf8
from .fable_modules.fable_library.option import (default_arg, map)
from .fable_modules.fable_library.reflection import (TypeInfo, string_type, union_type)
from .fable_modules.fable_library.string_ import (to_text, printf)
from .fable_modules.fable_library.types import (uint32, uint8, Array, Union)
from .binary import (read_uint32little_endian, read_byte, read_int32little_endian, write_uint32little_endian, write_byte, write_int32little_endian)

_T = TypeVar("_T")

def TryParseRequestHeader(buffer: bytearray) -> RequestPacket | None:
    if len(buffer) < 18:
        return None

    else: 
        try: 
            magic_number: uint32 = read_uint32little_endian(buffer, 0)
            version: uint8 = read_byte(buffer, 4)
            return RequestPacket(magic_number, version, int(read_byte(buffer, 5)), read_int32little_endian(buffer, 6), read_int32little_endian(buffer, 10), read_int32little_endian(buffer, 14), bytearray([]), None) if ((version == uint8(1)) if (magic_number == uint32(1397310287)) else False) else None

        except Exception as match_value:
            return None




def TryParseRequestPacket(buffer: bytearray) -> RequestPacket | None:
    match_value: RequestPacket | None = TryParseRequestHeader(buffer)
    if match_value is None:
        return None

    else: 
        header: RequestPacket = match_value
        if len(buffer) >= header.TotalPacketLength:
            try: 
                return RequestPacket(header.MagicNumber, header.Version, header.CommandType, header.TotalPacketLength, header.KeyLength, header.ValueLength, get_sub_array(buffer, 18, header.KeyLength), get_sub_array(buffer, 18 + header.KeyLength, header.ValueLength) if (header.ValueLength > 0) else None)

            except Exception as match_value_1:
                return None


        else: 
            return None




def SerializeRequest(packet: RequestPacket) -> bytearray:
    key_len: int = len(packet.Key) or 0
    def mapping(v: bytearray, packet: Any=packet) -> int:
        return len(v)

    buffer: bytearray = [0] * ((18 + key_len) + default_arg(map(mapping, packet.Value), 0))
    offset: int = 0
    write_uint32little_endian(packet.MagicNumber, buffer, offset)
    offset = (offset + 4) or 0
    write_byte(packet.Version, buffer, offset)
    offset = (offset + 1) or 0
    write_byte(int(packet.CommandType+0x100 if packet.CommandType < 0 else packet.CommandType) & 0xFF, buffer, offset)
    offset = (offset + 1) or 0
    write_int32little_endian(packet.TotalPacketLength, buffer, offset)
    offset = (offset + 4) or 0
    write_int32little_endian(packet.KeyLength, buffer, offset)
    offset = (offset + 4) or 0
    write_int32little_endian(packet.ValueLength, buffer, offset)
    offset = (offset + 4) or 0
    if key_len > 0:
        copy_to(packet.Key, 0, buffer, offset, key_len)
        offset = (offset + key_len) or 0

    match_value: bytearray | None = packet.Value
    if match_value is None:
        pass

    else: 
        value_1: bytearray = match_value
        copy_to(value_1, 0, buffer, offset, len(value_1))

    return buffer


def CreateSuccessResponse(value: bytearray | None=None) -> ResponsePacket:
    pattern_input: tuple[int, int]
    if value is None:
        pattern_input = (0, 18)

    else: 
        v: bytearray = value
        pattern_input = (len(v), 18 + len(v))

    return ResponsePacket(uint32(1397310287), uint8(1), uint8(0), pattern_input[1], pattern_input[0], 0, value, None)


def CreateNotFoundResponse(__unit: None=None) -> ResponsePacket:
    return ResponsePacket(uint32(1397310287), uint8(1), uint8(1), 18, 0, 0, None, None)


def CreateErrorResponse(error_message: str) -> ResponsePacket:
    msg_bytes: bytearray = get_utf8().get_bytes(error_message)
    return ResponsePacket(uint32(1397310287), uint8(1), uint8(2), 18 + len(msg_bytes), 0, len(msg_bytes), None, error_message)


def IsValidPacketSize(total_length: int) -> bool:
    if total_length >= 18:
        return total_length <= ((10 * 1024) * 1024)

    else: 
        return False



def SerializeResponse(packet: ResponsePacket) -> bytearray:
    def mapping(v: bytearray, packet: Any=packet) -> int:
        return len(v)

    def mapping_1(m: str, packet: Any=packet) -> int:
        return len(get_utf8().get_bytes(m))

    buffer: bytearray = [0] * ((18 + default_arg(map(mapping, packet.Value), 0)) + default_arg(map(mapping_1, packet.ErrorMessage), 0))
    offset: int = 0
    write_uint32little_endian(packet.MagicNumber, buffer, offset)
    offset = (offset + 4) or 0
    write_byte(packet.Version, buffer, offset)
    offset = (offset + 1) or 0
    write_byte(packet.StatusCode, buffer, offset)
    offset = (offset + 1) or 0
    write_int32little_endian(packet.TotalPacketLength, buffer, offset)
    offset = (offset + 4) or 0
    write_int32little_endian(packet.ValueLength, buffer, offset)
    offset = (offset + 4) or 0
    write_int32little_endian(packet.ErrorMessageLength, buffer, offset)
    offset = (offset + 4) or 0
    match_value: bytearray | None = packet.Value
    if match_value is None:
        pass

    else: 
        value_2: bytearray = match_value
        copy_to(value_2, 0, buffer, offset, len(value_2))
        offset = (offset + len(value_2)) or 0

    match_value_1: str | None = packet.ErrorMessage
    if match_value_1 is None:
        pass

    else: 
        msg: str = match_value_1
        msg_bytes: bytearray = get_utf8().get_bytes(msg)
        copy_to(msg_bytes, 0, buffer, offset, len(msg_bytes))

    return buffer


def _expr3(gen0: TypeInfo) -> TypeInfo:
    return union_type("Ocis.Client.SDK.Protocol.ParseResult`1", [gen0], ParseResult_1, lambda: [[("Item", gen0)], [("Item", string_type)], []])


class ParseResult_1(Union, Generic[_T]):
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["ParseSuccess", "ParseError", "InsufficientData"]


ParseResult_1_reflection = _expr3

def DeserializeResponse(buffer: bytearray) -> ParseResult_1[ResponsePacket]:
    try: 
        if len(buffer) < 18:
            return ParseResult_1(2)

        else: 
            offset: int = 0
            magic_number: uint32 = read_uint32little_endian(buffer, offset)
            offset = (offset + 4) or 0
            version: uint8 = read_byte(buffer, offset)
            offset = (offset + 1) or 0
            status_code_byte: uint8 = read_byte(buffer, offset)
            offset = (offset + 1) or 0
            total_packet_length: int = read_int32little_endian(buffer, offset) or 0
            offset = (offset + 4) or 0
            value_length: int = read_int32little_endian(buffer, offset) or 0
            offset = (offset + 4) or 0
            error_message_length: int = read_int32little_endian(buffer, offset) or 0
            offset = (offset + 4) or 0
            if len(buffer) < total_packet_length:
                return ParseResult_1(2)

            elif not ((version == uint8(1)) if (magic_number == uint32(1397310287)) else False):
                return ParseResult_1(1, "Invalid header")

            elif True if (value_length < 0) else (error_message_length < 0):
                return ParseResult_1(1, "Invalid length field")

            elif total_packet_length != ((18 + value_length) + error_message_length):
                return ParseResult_1(1, "Packet length mismatch")

            else: 
                value: bytearray | None = get_sub_array(buffer, offset, value_length) if (value_length > 0) else None
                offset = (offset + value_length) or 0
                def _arrow5(__unit: None=None) -> str | None:
                    error_bytes: bytearray = get_sub_array(buffer, offset, error_message_length)
                    return get_utf8().get_string(error_bytes)

                return ParseResult_1(0, ResponsePacket(magic_number, version, status_code_byte, total_packet_length, value_length, error_message_length, value, _arrow5() if (error_message_length > 0) else None))



    except Exception as ex:
        def _arrow4(__unit: None=None) -> str:
            arg: str = str(ex)
            return to_text(printf("Error parsing response: %s"))(arg)

        return ParseResult_1(1, _arrow4())



__all__ = ["TryParseRequestHeader", "TryParseRequestPacket", "SerializeRequest", "CreateSuccessResponse", "CreateNotFoundResponse", "CreateErrorResponse", "IsValidPacketSize", "SerializeResponse", "ParseResult_1_reflection", "DeserializeResponse"]

File: Ocis.Client/sdk/ts/Protocol.ts
import { writeInt32LittleEndian, writeByte, writeUInt32LittleEndian, readInt32LittleEndian, readByte, readUInt32LittleEndian } from "./Binary.js";
import { int32, uint8, uint32 } from "./fable_modules/fable-library-ts.4.25.0/Int32.js";
import { ResponsePacket, RequestPacket } from "./Ocis.Server/ProtocolSpec.js";
import { map, defaultArg, value as value_3, Option } from "./fable_modules/fable-library-ts.4.25.0/Option.js";
import { copyTo, fill, getSubArray } from "./fable_modules/fable-library-ts.4.25.0/Array.js";
import { get_UTF8 } from "./fable_modules/fable-library-ts.4.25.0/Encoding.js";
import { Union } from "./fable_modules/fable-library-ts.4.25.0/Types.js";
import { union_type, string_type, TypeInfo } from "./fable_modules/fable-library-ts.4.25.0/Reflection.js";
import { printf, toText } from "./fable_modules/fable-library-ts.4.25.0/String.js";

export function TryParseRequestHeader(buffer: uint8[]): Option<RequestPacket> {
    if (buffer.length < 18) {
        return undefined;
    }
    else {
        try {
            const magicNumber: uint32 = readUInt32LittleEndian(buffer, 0);
            const version: uint8 = readByte(buffer, 4);
            return ((magicNumber === 1397310287) && (version === 1)) ? (new RequestPacket(magicNumber, version, ~~readByte(buffer, 5), readInt32LittleEndian(buffer, 6), readInt32LittleEndian(buffer, 10), readInt32LittleEndian(buffer, 14), [], undefined)) : undefined;
        }
        catch (matchValue: any) {
            return undefined;
        }
    }
}

export function TryParseRequestPacket(buffer: uint8[]): Option<RequestPacket> {
    const matchValue: Option<RequestPacket> = TryParseRequestHeader(buffer);
    if (matchValue == null) {
        return undefined;
    }
    else {
        const header: RequestPacket = value_3(matchValue);
        if (buffer.length >= header.TotalPacketLength) {
            try {
                return new RequestPacket(header.MagicNumber, header.Version, header.CommandType, header.TotalPacketLength, header.KeyLength, header.ValueLength, getSubArray<uint8>(buffer, 18, header.KeyLength), (header.ValueLength > 0) ? getSubArray<uint8>(buffer, 18 + header.KeyLength, header.ValueLength) : undefined);
            }
            catch (matchValue_1: any) {
                return undefined;
            }
        }
        else {
            return undefined;
        }
    }
}

export function SerializeRequest(packet: RequestPacket): uint8[] {
    const keyLen: int32 = packet.Key.length | 0;
    const totalLen: int32 = ((18 + keyLen) + defaultArg(map<uint8[], int32>((v: uint8[]): int32 => v.length, packet.Value), 0)) | 0;
    const buffer: uint8[] = fill(new Array(totalLen), 0, totalLen, 0);
    let offset = 0;
    writeUInt32LittleEndian(packet.MagicNumber, buffer, offset);
    offset = ((offset + 4) | 0);
    writeByte(packet.Version, buffer, offset);
    offset = ((offset + 1) | 0);
    writeByte(packet.CommandType & 0xFF, buffer, offset);
    offset = ((offset + 1) | 0);
    writeInt32LittleEndian(packet.TotalPacketLength, buffer, offset);
    offset = ((offset + 4) | 0);
    writeInt32LittleEndian(packet.KeyLength, buffer, offset);
    offset = ((offset + 4) | 0);
    writeInt32LittleEndian(packet.ValueLength, buffer, offset);
    offset = ((offset + 4) | 0);
    if (keyLen > 0) {
        copyTo<uint8>(packet.Key, 0, buffer, offset, keyLen);
        offset = ((offset + keyLen) | 0);
    }
    const matchValue: Option<uint8[]> = packet.Value;
    if (matchValue == null) {
    }
    else {
        const value_1: uint8[] = value_3(matchValue);
        copyTo<uint8>(value_1, 0, buffer, offset, value_1.length);
    }
    return buffer;
}

export function CreateSuccessResponse(value: Option<uint8[]>): ResponsePacket {
    let patternInput: [int32, int32];
    if (value == null) {
        patternInput = ([0, 18] as [int32, int32]);
    }
    else {
        const v: uint8[] = value_3(value);
        patternInput = ([v.length, 18 + v.length] as [int32, int32]);
    }
    return new ResponsePacket(1397310287, 1, 0, patternInput[1], patternInput[0], 0, value, undefined);
}

export function CreateNotFoundResponse(): ResponsePacket {
    return new ResponsePacket(1397310287, 1, 1, 18, 0, 0, undefined, undefined);
}

export function CreateErrorResponse(errorMessage: string): ResponsePacket {
    const msgBytes: uint8[] = Array.from(get_UTF8().getBytes(errorMessage));
    return new ResponsePacket(1397310287, 1, 2, 18 + msgBytes.length, 0, msgBytes.length, undefined, errorMessage);
}

export function IsValidPacketSize(totalLength: int32): boolean {
    if (totalLength >= 18) {
        return totalLength <= ((10 * 1024) * 1024);
    }
    else {
        return false;
    }
}

export function SerializeResponse(packet: ResponsePacket): uint8[] {
    const totalLen: int32 = ((18 + defaultArg(map<uint8[], int32>((v: uint8[]): int32 => v.length, packet.Value), 0)) + defaultArg(map<string, int32>((m: string): int32 => Array.from(get_UTF8().getBytes(m)).length, packet.ErrorMessage), 0)) | 0;
    const buffer: uint8[] = fill(new Array(totalLen), 0, totalLen, 0);
    let offset = 0;
    writeUInt32LittleEndian(packet.MagicNumber, buffer, offset);
    offset = ((offset + 4) | 0);
    writeByte(packet.Version, buffer, offset);
    offset = ((offset + 1) | 0);
    writeByte(packet.StatusCode, buffer, offset);
    offset = ((offset + 1) | 0);
    writeInt32LittleEndian(packet.TotalPacketLength, buffer, offset);
    offset = ((offset + 4) | 0);
    writeInt32LittleEndian(packet.ValueLength, buffer, offset);
    offset = ((offset + 4) | 0);
    writeInt32LittleEndian(packet.ErrorMessageLength, buffer, offset);
    offset = ((offset + 4) | 0);
    const matchValue: Option<uint8[]> = packet.Value;
    if (matchValue == null) {
    }
    else {
        const value_2: uint8[] = value_3(matchValue);
        copyTo<uint8>(value_2, 0, buffer, offset, value_2.length);
        offset = ((offset + value_2.length) | 0);
    }
    const matchValue_1: Option<string> = packet.ErrorMessage;
    if (matchValue_1 == null) {
    }
    else {
        const msg: string = value_3(matchValue_1);
        const msgBytes: uint8[] = Array.from(get_UTF8().getBytes(msg));
        copyTo<uint8>(msgBytes, 0, buffer, offset, msgBytes.length);
    }
    return buffer;
}

export type ParseResult$1_$union<T> = 
    | ParseResult$1<T, 0>
    | ParseResult$1<T, 1>
    | ParseResult$1<T, 2>

export type ParseResult$1_$cases<T> = {
    0: ["ParseSuccess", [T]],
    1: ["ParseError", [string]],
    2: ["InsufficientData", []]
}

export function ParseResult$1_ParseSuccess<T>(Item: T) {
    return new ParseResult$1<T, 0>(0, [Item]);
}

export function ParseResult$1_ParseError<T>(Item: string) {
    return new ParseResult$1<T, 1>(1, [Item]);
}

export function ParseResult$1_InsufficientData<T>() {
    return new ParseResult$1<T, 2>(2, []);
}

export class ParseResult$1<T, Tag extends keyof ParseResult$1_$cases<T>> extends Union<Tag, ParseResult$1_$cases<T>[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: ParseResult$1_$cases<T>[Tag][1]) {
        super();
    }
    cases() {
        return ["ParseSuccess", "ParseError", "InsufficientData"];
    }
}

export function ParseResult$1_$reflection(gen0: TypeInfo): TypeInfo {
    return union_type("Ocis.Client.SDK.Protocol.ParseResult`1", [gen0], ParseResult$1, () => [[["Item", gen0]], [["Item", string_type]], []]);
}

export function DeserializeResponse(buffer: uint8[]): ParseResult$1_$union<ResponsePacket> {
    let arg: string, errorBytes: uint8[];
    try {
        if (buffer.length < 18) {
            return ParseResult$1_InsufficientData<ResponsePacket>();
        }
        else {
            let offset = 0;
            const magicNumber: uint32 = readUInt32LittleEndian(buffer, offset);
            offset = ((offset + 4) | 0);
            const version: uint8 = readByte(buffer, offset);
            offset = ((offset + 1) | 0);
            const statusCode: uint8 = readByte(buffer, offset);
            offset = ((offset + 1) | 0);
            const totalPacketLength: int32 = readInt32LittleEndian(buffer, offset) | 0;
            offset = ((offset + 4) | 0);
            const valueLength: int32 = readInt32LittleEndian(buffer, offset) | 0;
            offset = ((offset + 4) | 0);
            const errorMessageLength: int32 = readInt32LittleEndian(buffer, offset) | 0;
            offset = ((offset + 4) | 0);
            if (buffer.length < totalPacketLength) {
                return ParseResult$1_InsufficientData<ResponsePacket>();
            }
            else if (!((magicNumber === 1397310287) && (version === 1))) {
                return ParseResult$1_ParseError<ResponsePacket>("Invalid header");
            }
            else if ((valueLength < 0) ? true : (errorMessageLength < 0)) {
                return ParseResult$1_ParseError<ResponsePacket>("Invalid length field");
            }
            else if (totalPacketLength !== ((18 + valueLength) + errorMessageLength)) {
                return ParseResult$1_ParseError<ResponsePacket>("Packet length mismatch");
            }
            else {
                const value: Option<uint8[]> = (valueLength > 0) ? getSubArray<uint8>(buffer, offset, valueLength) : undefined;
                offset = ((offset + valueLength) | 0);
                return ParseResult$1_ParseSuccess<ResponsePacket>(new ResponsePacket(magicNumber, version, statusCode, totalPacketLength, valueLength, errorMessageLength, value, (errorMessageLength > 0) ? ((errorBytes = getSubArray<uint8>(buffer, offset, errorMessageLength), get_UTF8().getString(errorBytes))) : undefined));
            }
        }
    }
    catch (ex: any) {
        return ParseResult$1_ParseError<ResponsePacket>((arg = ex.message, toText(printf("Error parsing response: %s"))(arg)));
    }
}

